astrolabe/utils.py

```diff
@@ -0,0 +1,132 @@
+from collections import defaultdict
+import errno
+import json
+import os
+from textwrap import dedent
+from time import monotonic
+
+import click
+import junitparser
+
+
+def assert_subset(dict1, dict2):
+    """Utility that asserts that `dict2` is a subset of `dict1`, while
+    accounting for nested fields."""
+    for key, value in dict2.items():
+        if key not in dict1:
+            raise AssertionError("not a subset")
+        if isinstance(value, dict):
+            assert_subset(dict1[key], value)
+        else:
+            assert dict1[key] == value
+
+
+class Timer:
+    """Class to simplify timing operations."""
+    def __init__(self):
+        self._start = None
+        self._end = None
+
+    def reset(self):
+        self.__init__()
+
+    def start(self):
+        self._start = monotonic()
+        self._end = None
+
+    def stop(self):
+        self._end = monotonic()
+
+    @property
+    def elapsed(self):
+        if self._end is None:
+            return monotonic() - self._start
+        return self._end - self._start
+
+
+def cached_property(func):
+    """Decorator to memoize a class method that accepts no args/kwargs."""
+    memo = None
+
+    def memoized_function(self, *args, **kwargs):
+        if args or kwargs:
+            raise RuntimeError("cannot memoize methods that accept arguments")
+        nonlocal memo
+        if memo is None:
+            memo = func(self)
+        return memo
+
+    return memoized_function
+
+
+def encode_cdata(data):
+    """Encode `data` to XML-recognized CDATA."""
+    return "<![CDATA[{data}]]>".format(data=data)
+
+
+class SingleTestXUnitLogger:
+    def __init__(self, *, output_directory):
+        self._output_directory = os.path.realpath(os.path.join(
+            os.getcwd(), output_directory))
+
+        # Ensure folder exists.
+        try:
+            os.mkdir(self._output_directory)
+        except FileExistsError:
+            pass
+
+    def write_xml(self, test_case, filename):
+        filename += '.xml'
+        xml_path = os.path.join(self._output_directory, filename)
+
+        # Remove existing file if applicable.
+        try:
+            os.unlink(xml_path)
+        except FileNotFoundError:
+            pass
+
+        # use filename as suitename
+        suite = junitparser.TestSuite(filename)
+        suite.add_testcase(test_case)
+
+        xml = junitparser.JUnitXml()
+        xml.add_testsuite(suite)
+        xml.write(xml_path)
+
+
+def _nested_defaultdict():
+    """An infinitely nested defaultdict type."""
+    return defaultdict(_nested_defaultdict)
+
+
+def _merge_dictionaries(dicts):
+    """Utility method to merge a list of dictionaries.
+    Last observed value prevails."""
+    result = {}
+    for d in dicts:
+        result.update(d)
+    return result
+
+
+class _JsonDotNotationType(click.ParamType):
```

Comment:
A unit test for this would be nice. Or at least some sample input/output in the docstring.

Reply:
I can add a unittest but I am not so sure we need to support free-form text input of JSON-fields (which is what this class was designed for) anymore which might mean we can get rid of this altogether. Fleshing out the last remaining CLI endpoints will make things clearer.

Reply:
This is no longer needed.